accepted/future-releases/constructor-tearoffs/faq.md

```diff
@@ -0,0 +1,256 @@
+
+Examples:
+
+```dart
+Type intListType = List<int>; // Explicit type literal instantiation.
+
+T id<T>(T value) => value; // Our standard generic function example.
+var idValue = id; // A function *value*.
+
+var intId = id<int>; // Explicit instantiation, saves on writing the function type.
+var intId2 = idValue<int>; // Still works!
+var intId3 = (id)<int>; // Still works!
+
+var makeList = (List.filled)<String>; // List<String> Function(int, String)
+```
+
+The last example shows the generality of the feature, because the `(List.filled)` tear-off is a generic function value, you can instantiate it. *Don't even write that*, always use `List<String>.filled` instead!
+
+In short, we allow you to use `<typeArguments>` as a *selector*, like `.name` and `[expr]`, which you can chain after an expression. Previously we only allowed type arguments to occur before an argument list (`<typeArguments>(arguments)`) or after a class name in a constructor invocation (`ClassName<typeArguments>.name(arguments)`). Now we allow it after any expression, and with the same precedence as other selectors like argument lists, `.name` and `[expr]`.
+
+Whenever such a type argument list is followed by an argument list, it exactly means the same as it used to. No change there.
+
+**Q:** Doesn't that make the grammar, like, totally ambiguous?
+
+**A:** Yes! *Thank you* for noticing! And that is a problem. With Dart 2.0 we introduced generic function invocations, and had to decide how to parse `f(a<b,c>(d))`. The argument(s) to `f` could be either two comparison operator expressions or a single generic function invocation. We decided on always choosing the latter when `b` and `c` can be parsed as types and the `>` is followed by a `(`. (We have to decide while we parse the program, long before we can even begin to figure out what `b` and `c` actually refer to, so the choice is entirely grammar based.) We now have even more similar ambiguous cases. For example `f(a<b,c>-d)` is ambiguous because `-` can both be a prefix operator after a greater-than operator, or an infix operator after an explicit type-instantiation. Our choice is to be very restrictive in when we parse `expr <` as starting a type argument. We only do so when the following tokens *can* be parsed as a type argument list, and the only if the *next token* after the final `>` of the type arguments is one of:
+
+> `)`, `}`, `]`, `;`, `:`, `,`,`(`, `.`, `==`, or `!=`
+
+If the next token is *any other token*, then the `<` is parsed as a less-than operator.
```

Comment:
In case people are worried they're going to have to memorize this rule, it might be worth adding a final sentence saying something like:

> In practice, we believe the parser will just do the right thing without any effort on your part.  In the unlikely event that it doesn't, you can force the interpretation you want by adding parentheses.

Reply:
Done.